src/transformers/modeling_tf_utils.py

```diff
@@ -47,6 +47,36 @@ def num_parameters(self, only_trainable: bool = False) -> int:
             return self.count_params()
 
 
+def keras_serializable(cls):
+    initializer = cls.__init__
+
+    def wrapped_init(self, config, *args, **kwargs):
+        if isinstance(config, dict):
+            from transformers import AutoConfig
+
+            config = AutoConfig.config_class_for_model_class(cls).from_dict(config)
+        initializer(self, config, *args, **kwargs)
+        self._transformers_config = config
+
+    cls.__init__ = wrapped_init
+
+    if not hasattr(cls, "get_config"):
+        raise TypeError("Only use @keras_serializable on tf.keras.layers.Layer subclasses")
+    if hasattr(cls.get_config, "_is_default"):
+
+        def get_config(self):
+            cfg = super(cls, self).get_config()
+            cfg["config"] = self._transformers_config.to_dict()
+            return cfg
+
+        cls.get_config = get_config
+
+    cls._keras_serializable = True
+    if hasattr(tf.keras.utils, "register_keras_serializable"):
+        cls = tf.keras.utils.register_keras_serializable()(cls)
+    return cls
```

Comment:
I'm not sure about the choice of a wrapper instead of just inheriting from a base `TFMainLayer` like you mentioned in #3101.  I feel like a base class would be more easy to read and understand rather than this complex scheme of multiple wrapping.

Reply:
See rationale in https://github.com/huggingface/transformers/pull/3103/commits/0c716ede8c654cd2cdd3ab924f099e5d8277ee68 — I agree that a `TFMainLayer` superclass would be simpler and I tried that first (see first commit in this PR) but soon realized that there's no way to make Keras deserialization work that way, because Keras will pass the JSON-deserialized object (typically a dict) to the initializer — but the initializers of layers in this project expect an instance of `PretrainedConfig` or a subclass of it. A superclass can't affect what arguments are passed to a subclass initializer. But a class decorator can. So that's what I fell back on.

An alternative would be to add (nearly identical) code to _all_ those initializers to handle the dict case as well as the config object case (and document a rule that new main-layer classes should do that also). That seemed messier and less simple than the class decorator approach.

Reply:
> rather than this complex scheme of multiple wrapping.

Note, there isn't actually any _multiple_ wrapping — the only thing that gets _wrapped_ is the `__init__` method (to intercept the `config` parameter). The thing that gets returned is the original `cls` object, just with these modifications applied to it:

* wrapping `__init__` (the reason this can't just be a superclass)
* overriding `get_config` to add the `config` attribute
* adding the `_keras_serializable = True` attribute so it can be checked for (currently only used by the test, to decide which layer classes to test)

(Also calls `register_keras_serializable()(cls)` if available, but that just inspects the class and stores references to it, but returns the class itself unmodified. So in principle we could just call `register_keras_serializable()` it and not assign its return value to `cls`, but that would be baking knowledge of its current implementation into this code.)

Reply:
Ok got it, thanks for the details.

I like your approach!